### ISSUE TITLE
ReadOnlySpan<char>.Trim for small inputs is few times slower on Linux

### ISSUE THREAD

**Issue by adamsitnik** [dotnet/runtime#13669]
`ReadOnlySpan<char>.Trim` for small inputs is few times slower on Linux

| Slower                                                          | Lin/Win | Win Median (ns) | Lin Median (ns) | Modality|
| --------------------------------------------------------------- | -------:| ---------------:| ---------------:| --------:|
| System.Memory.ReadOnlySpan.Trim(input: "")                      |    6.45 |            3.36 |           21.68 |         |
| System.Memory.ReadOnlySpan.Trim(input: "abcdefg")               |    5.67 |            4.82 |           27.33 |         |
| System.Memory.ReadOnlySpan.Trim(input: " abcdefg ")             |    3.32 |            7.78 |           25.83 |         |

Benchmark:

https://github.com/dotnet/performance/blob/8b23cabe793b4ff73a9b28c7dd092b11dc17b197/src/benchmarks/micro/corefx/System.Memory/ReadOnlySpan.cs#L77-L79

```cmd
git clone https://github.com/dotnet/performance.git
python3 ./performance/scripts/benchmarks_ci.py -f netcoreapp5.0 --filter System.Memory.ReadOnlySpan.Trim
```

I've created a very small repro app:

```cs
class Program
{
    static int Main(string[] args)
    {
        int result = 0;

        ReadOnlySpan<char> span = string.Empty.AsSpan();
        for (int i = 0; i < 1_000_000_000; i++)
        {
            result ^= TrimSourceCopied(span).Length;
        }

        return result;
    }
    
    private static ReadOnlySpan<char> TrimSourceCopied(ReadOnlySpan<char> span)
    {
        int start = 0;
        for (; start < span.Length; start++)
        {
            if (!char.IsWhiteSpace(span[start]))
            {
                break;
            }
        }

        int end = span.Length - 1;
        for (; end > start; end--)
        {
            if (!char.IsWhiteSpace(span[end]))
            {
                break;
            }
        }

        return span.Slice(start, end - start + 1);
    }
}
```

And using VTune I was able to narrow down the problem to struct copying:

The body of the `Main` method on Ubuntu 18.04:

![image](https://user-images.githubusercontent.com/6011991/67693093-6f11d000-f9a1-11e9-9436-7153aad8d9ee.png)

Please mind the body of the loop:

![image](https://user-images.githubusercontent.com/6011991/67692592-a9c73880-f9a0-11e9-9552-24ea267adc0e.png)

When I change `TrimSourceCopied` to accept the `Span` as `readonly ref` parameter:

```cs
private static ReadOnlySpan<char> TrimSourceCopied(in ReadOnlySpan<char> span)
```

![image](https://user-images.githubusercontent.com/6011991/67693428-f5c6ad00-f9a1-11e9-9d78-ec883df3323b.png)

`System.Memory.ReadOnlySpan.Trim` is only one example of a benchmark that uses `Span` a lot and is slower on Linux compared to Windows. This pattern|problem is quite common.

/cc @AndyAyersMS 

category:cq
theme:structs
skill-level:expert
cost:medium

**Comment by CarolEidt** [comment]
The struct values passed to `TrimSourceCopied` are now kept in registers, but more work is needed to ensure that they remain in registers in the called method.

**Comment by CarolEidt** [comment]
Fixed by #43870

**Comment by adamsitnik** [comment]
@CarolEidt awesome! can't wait to see how many benchmarks show improvement on Linux

FWIW these 3 charts should show it for this particular method in the next 24 hours: [1](https://pvscmdupload.blob.core.windows.net/reports/allTestHistory%2frefs%2fheads%2fmaster_x64_ubuntu%2018.04%2fSystem.Memory.ReadOnlySpan.Trim(input%3a%20%22%22).html), [2](https://pvscmdupload.blob.core.windows.net/reports/allTestHistory%2frefs%2fheads%2fmaster_x64_ubuntu%2018.04%2fSystem.Memory.ReadOnlySpan.Trim(input%3a%20%22abcdefg%22).html), [3](https://pvscmdupload.blob.core.windows.net/reports/allTestHistory%2frefs%2fheads%2fmaster_x64_ubuntu%2018.04%2fSystem.Memory.ReadOnlySpan.Trim(input%3a%20%22%20abcdefg%20%22).html)
